packages/multichain-account-api/src/adapters.ts

```diff
@@ -0,0 +1,292 @@
+    this.#accounts.set(groupIndex, multichainAccount);
+
+    return multichainAccount;
+  }
+
+  getNextGroupIndex(): number {
+    // Assuming we cannot have indexes gaps.
+    return this.#accounts.size; // No +1 here, group indexes starts at 0.
+  }
+
+  async createMultichainAccount(
+    groupIndex: number,
+  ): Promise<MultichainAccount<Account>> {
+    const nextGroupIndex = this.getNextGroupIndex();
+    if (groupIndex > nextGroupIndex) {
+      throw new Error(
+        `You cannot use an group index that is higher than the next available one: expect <${nextGroupIndex}, got ${groupIndex}`,
+      );
+    }
+
+    for (const provider of this.#providers) {
+      await provider.createAccounts({
+        entropySource: this.#entropySource,
+        groupIndex,
+      });
+    }
+
+    // Re-create and "refresh" the multichain account (we assume all account creations are
+    // idempotent, so we should get the same accounts and potentially some new accounts (if
+    // some account providers decide to return more of them this time).
+    return this.#createMultichainAccount(groupIndex);
+  }
+
+  async createNextMultichainAccount(): Promise<MultichainAccount<Account>> {
+    return this.createMultichainAccount(this.getNextGroupIndex());
+  }
+
+  async discoverAndCreateMultichainAccounts(): Promise<
+    MultichainAccount<Account>[]
+  > {
+    const multichainAccounts: MultichainAccount<Account>[] = [];
+
+    let accounts: AccountId[];
+    let groupIndex = 0;
+
+    do {
+      // New index means new accounts.
+      accounts = [];
+
+      const missingProviders = [];
+      for (const provider of this.#providers) {
+        const discoveredAccounts = await provider.discoverAndCreateAccounts({
+          entropySource: this.#entropySource,
+          groupIndex,
+        });
+
+        if (discoveredAccounts.length) {
+          // This provider has discovered and created accounts, meaning there might
+          // be something to discover on the next index.
+          accounts = discoveredAccounts;
+        } else {
+          // This provider did not discover or create any accounts. We mark it as
+          // "missing", so we can create accounts on this index if other providers
+          // did discover something.
+          missingProviders.push(provider);
+        }
+      }
+
+      if (accounts.length) {
+        // We only create missing accounts if one of the provider has discovered
+        // and created accounts.
+        for (const provider of missingProviders) {
+          await provider.createAccounts({
+            entropySource: this.#entropySource,
+            groupIndex,
+          });
+        }
+
+        // We've got all the accounts now, we can create our multichain account.
+        multichainAccounts.push(this.#createMultichainAccount(groupIndex));
+
+        // We have accounts, we need to check the next index.
+        groupIndex += 1;
+      }
+    } while (accounts.length);
```

Comment:
Question: Maybe we should just run discovery with the providers that HAS discovered something? (e.g Provider1 found something on index 0, but Provider2 did not, then we continue with index 1 for Provider1 but without Provider2?

If we do that, then we can use the "alignment mechanism" at the end of the discovery to re-align all accounts accross the multichain account.